Source/WebCore/platform/mac/ThemeMac.mm

```diff
@@ -493,61 +493,6 @@ static void setUpButtonCell(NSButtonCell *cell, ControlPartType type, const Cont
     setUpButtonCell(cell, type, controlStates, zoomedSize, zoomFactor);
     return cell;
 }
-    
-static void paintButton(ControlPartType type, ControlStates& controlStates, GraphicsContext& context, const FloatRect& zoomedRect, float zoomFactor, ScrollView* scrollView, float deviceScaleFactor)
-{
-    BEGIN_BLOCK_OBJC_EXCEPTIONS
-    
-    // Determine the width and height needed for the control and prepare the cell for painting.
-    auto states = controlStates.states();
-    NSButtonCell *buttonCell = button(type, controlStates, IntSize(zoomedRect.size()), zoomFactor);
-    GraphicsContextStateSaver stateSaver(context);
-
-    NSControlSize controlSize = [buttonCell controlSize];
-    IntSize zoomedSize = buttonSizes()[controlSize];
-    zoomedSize.setWidth(zoomedRect.width()); // Buttons don't ever constrain width, so the zoomed width can just be honored.
-    zoomedSize.setHeight(zoomedSize.height() * zoomFactor);
-    FloatRect inflatedRect = zoomedRect;
-    if ([buttonCell bezelStyle] == NSBezelStyleRounded) {
-        // Center the button within the available space.
-        if (inflatedRect.height() > zoomedSize.height()) {
-            inflatedRect.setY(inflatedRect.y() + (inflatedRect.height() - zoomedSize.height()) / 2);
-            inflatedRect.setHeight(zoomedSize.height());
-        }
-
-        // Now inflate it to account for the shadow.
-        inflatedRect = inflateRect(inflatedRect, zoomedSize, buttonMargins(controlSize), zoomFactor);
-
-        if (zoomFactor != 1.0f) {
-            inflatedRect.setWidth(inflatedRect.width() / zoomFactor);
-            inflatedRect.setHeight(inflatedRect.height() / zoomFactor);
-            context.translate(inflatedRect.location());
-            context.scale(zoomFactor);
-            context.translate(-inflatedRect.location());
-        }
-    }
-    
-    LocalCurrentGraphicsContext localContext(context);
-    
-    NSView *view = ThemeMac::ensuredView(scrollView, controlStates);
-    NSWindow *window = [view window];
-    NSButtonCell *previousDefaultButtonCell = [window defaultButtonCell];
-
-    bool needsRepaint = ThemeMac::drawCellOrFocusRingWithViewIntoContext(buttonCell, context, inflatedRect, view, true, states.contains(ControlStates::States::Focused), deviceScaleFactor);
-    if (states.contains(ControlStates::States::Default))
-        [window setDefaultButtonCell:buttonCell];
```

Comment:
It doesn't look like the call to `setDefaultButtonCell` is preserved anywhere.

Reply:
I think the existing code is incorrect. It is supposed to set the default button cell before drawing and restore it after drawing it. This code used to look like this (See https://bugs.webkit.org/show_bug.cgi?id=21775)

```
    NSView *view = scrollView->documentView();
    NSWindow *window = [view window];
    NSButtonCell *previousDefaultButtonCell = [window defaultButtonCell];

    if ((states & DefaultState) && [window isKeyWindow]) {
        [window setDefaultButtonCell:buttonCell];
        wkAdvanceDefaultButtonPulseAnimation(buttonCell);
    } else if ([previousDefaultButtonCell isEqual:buttonCell])
        [window setDefaultButtonCell:nil];

    [buttonCell drawWithFrame:NSRect(inflatedRect) inView:view];
    [buttonCell setControlView:nil];

    if (![previousDefaultButtonCell isEqual:buttonCell])
        [window setDefaultButtonCell:previousDefaultButtonCell];

```
But this order was changed incorrectly in https://trac.webkit.org/changeset/187961/webkit. This change made the drawing happens before setting up and restoring the default button.

Reply:
I restored the order as it was before https://trac.webkit.org/changeset/187961/webkit.